core/templates/cowdata.h

```diff
@@ -272,139 +265,289 @@ void CowData<T>::_unref() {
 }
 
 template <typename T>
-Error CowData<T>::_fork_allocate(USize p_size) {
-	if (p_size == 0) {
-		// Wants to clean up.
+void CowData<T>::remove_at(Size p_index) {
+	const Size prev_size = size();
+	ERR_FAIL_INDEX(p_index, prev_size);
+
+	if (prev_size == 1) {
+		// Removing the only element.
 		_unref();
-		return OK;
+		return;
 	}
 
-	USize alloc_size;
-	ERR_FAIL_COND_V(!_get_alloc_size_checked(p_size, &alloc_size), ERR_OUT_OF_MEMORY);
+	const USize new_size = prev_size - 1;
 
-	const USize prev_size = size();
+	if (_get_refcount()->get() == 1) {
+		// We're the only owner; remove in-place.
 
-	if (!_ptr) {
-		// We had no data before; just allocate a new array.
-		const Error error = _alloc(alloc_size);
-		if (error) {
-			return error;
-		}
-	} else if (_get_refcount()->get() == 1) {
-		// Resize in-place.
-		// NOTE: This case is not just an optimization, but required, as some callers depend on
-		//       `_copy_on_write()` calls not changing the pointer after the first fork
-		//       (e.g. mutable iterators).
-		if (p_size == prev_size) {
-			// We can shortcut here; we don't need to do anything.
-			return OK;
-		}
+		// Destruct the element, then relocate the rest one down.
+		_ptr[p_index].~T();
+		memmove((void *)(_ptr + p_index), (void *)(_ptr + p_index + 1), (new_size - p_index) * sizeof(T));
 
-		// Destroy extraneous elements.
-		if constexpr (!std::is_trivially_destructible_v<T>) {
-			for (USize i = prev_size; i > p_size; i--) {
-				_ptr[i - 1].~T();
-			}
+		// Shrink buffer if necessary.
+		const USize new_alloc_size = _get_alloc_size(new_size);
+		const USize prev_alloc_size = _get_alloc_size(capacity());
+		if (new_alloc_size < prev_alloc_size) {
+			_ALLOW_DISCARD_ _realloc_bytes(new_alloc_size); // Always succeeds on shrink.
 		}
 
-		if (alloc_size != _get_alloc_size(prev_size)) {
-			const Error error = _realloc(alloc_size);
+		*_get_size() = new_size;
+	} else {
+		// Remove by forking.
+		CRASH_COND(_copy_to_new_buffer(new_size, p_index, 0, new_size - p_index));
```

Comment:
I would change to this, both because clearer, and also because it's possible someone will disable `CRASH_COND()` macro, in which case, it won't run at all.
```
Error err = _copy_to_new_buffer(new_size, p_index, 0, new_size - p_index);
CRASH_COND(err != OK);
```